loader.lua

```diff
@@ -0,0 +1,192 @@
+    self.len_num = 0 -- number of unique seqLengths
+
+end
+
+function indexer:prep_same_len_inds()
+    --[[
+        make a table of inds with ascending lens, so that we can return inds
+        with same seqLength using nxt_same_len_inds()
+    --]]
+
+    print('preparing the inds with the same seqLengths..')
+
+    self.db_spect:open(); local txn = self.db_spect:txn(true)
+    local len_set = {}
+    for i = 1, self.lmdb_size do
+        local _len = txn:get(i):size(2) -- get the len of spect
+        table.insert(self.same_len_inds, {i, _len})
+        if len_set[_len] == nil then len_set[_len] = true end
+        -- if i % 100 == 0 then xlua.progress(i, self.lmdb_size) end
+    end
+
+    txn:abort(); self.db_spect:close()
+
+    -- sort table
+    local function comp(a, b) return a[2] < b[2] end
+    table.sort(self.same_len_inds, comp)
+
+    --debug
+    --print(self.same_len_inds)
+
+    for _ in pairs(len_set) do self.len_num = self.len_num + 1 end -- number of different seqLengths
+    print('there are ' .. self.len_num .. ' unique seqLengths')
+end
+
+function indexer:nxt_same_len_inds()
+    --[[
+        return inds with same seqLength, a solution before zero-masking can work
+
+        NOTE:
+            call prep_same_len_inds before this
+    --]]
+    assert(#self.same_len_inds > 0, 'call prep_same_len_inds before this')
+
+    local _len = self.same_len_inds[self.cnt][2]
+    local inds = {}
+    while(self.cnt <= self.lmdb_size and self.same_len_inds[self.cnt][2] == _len) do
+        -- NOTE: true index store in table, instead of cnt
+        table.insert(inds, self.same_len_inds[self.cnt][1])
+        self.cnt = self.cnt + 1
+    end
+
+    if self.cnt > self.lmdb_size then self.cnt = 1 end
+
+    return inds
+end
+
+function indexer:nxt_inds()
+    --[[
+        return indexs of next batch
+    --]]
+
+    local inds = {}
+    if self.lmdb_size > self.cnt + self.batch_size - 1 then
+        for i=0,self.batch_size-1 do
+            table.insert(inds,self.cnt+i)
+        end
+
+        self.cnt = self.cnt + self.batch_size
+        return inds
+    end
+
+    -- case where cnt+size >= total size
+    for i=self.cnt,self.lmdb_size do
+        table.insert(inds,i)
+    end
+
+    self.cnt = self.batch_size - (self.lmdb_size - self.cnt)
+    for i=1, self.cnt-1 do -- overflow inds
+        table.insert(inds,i)
+    end
+
+    return inds
+
+end
+
+
+
+local loader = torch.class('loader')
+
+function loader:__init(_dir)
+    --[[
+        _dir: dir contains 3 lmdbs
+    --]]
+
+    self.db_spect = lmdb.env{Path=_dir..'/spect',Name='spect'}
+    self.db_label = lmdb.env{Path=_dir..'/label',Name='label'}
+    self.db_trans = lmdb.env{Path=_dir..'/trans',Name='trans'}
+end
+
+function loader:nxt_batch(inds, flag)
+    --[[
+        return a batch by loading from lmdb just-in-time
+
+        flag: indicates whether to load trans
+
+        TODO we allocate 2 * batch_size space
+    --]]
+    local tensor_list = {}
```

Comment:
I've had complaints of memory issues with storing it like this. What works is to use [this](https://github.com/torch/tds) and change this to local tensor_list = tds.Vec(). Also you have to import above tds = require 'tds'.

Also you would have to change table.insert(tensor_list, tensor) to tensor_list:insert(tensor). Everything else is good!